include/arch/arm/arch/64/mode/machine/debug.h

```diff
@@ -6,4 +6,141 @@
 
 #pragma once
 
+#ifdef ARM_BASE_CP14_SAVE_AND_RESTORE
 
+void aarch64_restore_user_debug_context(tcb_t *target_thread);
+
+/** Determines and carries out what needs to be done for a debug exception.
+ *
+ * This could be handling a single-stepping exception, or a breakpoint or
+ * watchpoint.
+ */
+seL4_Fault_t handleUserLevelDebugException(word_t esr, word_t fault_vaddr);
+bool_t isDebugFault(word_t esr);
+
+#define MAKE_DBGBVR(num) "DBGBVR" #num "_EL1"
+#define MAKE_DBGBCR(num) "DBGBCR" #num "_EL1"
+#define MAKE_DBGWVR(num) "DBGWVR" #num "_EL1"
+#define MAKE_DBGWCR(num) "DBGWCR" #num "_EL1"
+
+/** Generates read functions for the CP14 control and value registers.
+ */
+#define DEBUG_GENERATE_READ_FN(_name, _reg)                                    \
```

Comment:
I hate these macros and don't understand why they're needed. I honestly think writing all the used cases out would be less code and clearer than this indirect macro stuff.

That 32-bit debug code has this is no excuse to copy it, but if you do, at least share the macros: The only difference is MRS versus MRC, easily defined away...

Reply:
Yeah I don't really like the macros either, but I thought that duplicating them would at least make it slightly less horrible than adding another level of pre-processing in there with extra ifdefs and more macros.